### PR TITLE
  feat(ws1):Add kernel gtest operator checker and used logprob to test gtest

### PR DESCRIPTION
**Pull request by frank-2077**
## Summary
 - add 'rl_engine.kernels.gtest' ops correctness framework
 - add ops input generation
 - add 'scripts/check_operator.py'  CLI for comparing backend candidates against pytorch gold paths
 - add tolerance contract, contract loading, ops reports
part of #108 

## Usage

  Run a CPU smoke check against the PyTorch gold implementation:

```
  python scripts/check_operator.py \
    --op logp \
    --candidate pytorch \
    --device cpu \
    --dtype fp32 \
    --batch 1 \
    --seq 2 \
    --vocab 17

```
  Run a CUDA candidate check against the PyTorch gold path:

```
  python scripts/check_operator.py \
    --op logp \
    --candidate cuda \
    --device cuda \
    --dtype bf16 \
    --arch-key sm90 \
    --batch 1 \
    --seq 1 \
    --vocab 4096
```

  Print the full structured report as JSON:

```
  python scripts/check_operator.py \
    --op logp \
    --candidate pytorch \
    --device cpu \
    --dtype fp32 \
    --batch 1 \
    --seq 2 \
    --vocab 17 \
    --json

```
  Available key options:

  - --op: operator name. Current minimal version supports logp.
  - --candidate: backend candidate, for example pytorch, cuda, cuda-generic, cuda-sm90, or registry.
  - --dtype: fp32, bf16, or fp16.
  - --device: auto, cpu, cuda, or any torch device string.
  - --arch-key: optional tolerance override key, for example sm90.
  - --batch, --seq, --vocab: shape controls.
  - --input-mode: random or constant.
  - --constant-value: floating-point tensor value used by constant mode.
  - --token-value: token id used by constant mode, modulo vocab.
  - --seed: random input seed.
  - --json: emit full JSON report.

  Example output:

  suite=logp passed=True pass_rate=1.0000
  candidate=cuda-logp backend=cuda passed=True pass_rate=1.0000
    case=logp-torch.bfloat16-1x1x4096 output=0 shape=(1, 1) dtype=torch.bfloat16 max_abs=2.69813538e-02 mean_abs=2.69813538e-02 max_rel=3.03093810e-03
    tol=(atol=5.000e-02, rtol=0.000e+00) passed=True

## Adding a New Operator

  To add a new operator to the checker, keep the public test flow unchanged and only update the operator-specific registration/input files.

  ### 1. Add input generation
  **(Already added, need check the shapes)**
```
  rl_engine/kernels/gtest/operator_inputs.py 


  Update make_operator_inputs():

  builders = {
      ...
      "new_op": _make_new_op_inputs,
  }

  Update operator_shape_name():

  names = {
      ...
      "new_op": f"{batch}x{seq}x...",
  }

  Add the input builder:

  def _make_new_op_inputs(args, dtype, device):
      batch, seq = _batch_seq(args)
      return {
          "x": _floating_tensor((batch, seq, ...), args, dtype, device, offset=0),
          ...
      }

```

  ### 2. Register gold and candidates
**(NEED OPS OWNER ADD)**
```

  rl_engine/kernels/gtest/operator_specs.py

  Add an OperatorSpec entry:

  "new_op": OperatorSpec(
      name="new_op",
      op_class="elementwise",
      gold_path="rl_engine.kernels.ops.pytorch....NativeNewOp",
      registry_name="new_op",
      candidate_paths={
          "pytorch": "rl_engine.kernels.ops.pytorch....NativeNewOp",
          "cuda": "rl_engine.kernels.ops.cuda....CudaNewOp",
          "triton": "rl_engine.kernels.ops.triton....TritonNewOp",
      },
  )

```
  Rules:

  - gold_path must come from rl_engine.kernels.ops.pytorch.
  - CUDA/Triton/ROCm implementations are candidates only.
  - Do not compare two operators that implement different math.
  - candidate=pytorch is only for checker smoke tests.

## Validation
 - `python scripts/check_operator.py --op logp --candidate pytorch --device cpu --dtype fp32 --batch 1 --seq 2 --vocab 17`

71837019648a53174ec8e566ce210027

**In fact, the test did not pass; however, it proves that the workflow of our testing framework has achieved the minimum viable capability.**

## Notes

  - Gold paths are required to come from `rl_engine.kernels.ops.pytorch`.
  - CUDA/Triton/ROCm implementations are treated as candidates.
  - SM90 fused logp remains under separate validation and is not included as a passing path in this PR.

## Summary by CodeRabbit

* **New Features**
  * Added a command-line tool to compare operator results against a reference implementation, with support for dtype, device, input modes, seeds, and optional gradient checks.
  * Expanded operator coverage for `logp` and `linear_logp`, including standardized input generation and tolerance-based comparisons.
* **Bug Fixes**
  * Improved accuracy reporting for edge cases and added support for batch-invariance and backward-pass validation.
* **Documentation**
  * Updated operator guidance and contribution notes with usage examples, test commands, and validation steps.

### CODE REVIEW

File: docs/contributing/issue-108-session-log.md [hunk 1]
```diff
@@ -0,0 +1,1044 @@
+# ISSUE-108 Session Log
+
+本文档记录本 session 中围绕 RL-Kernel 算子测试框架、CUDA 验证和 upstream 同步的所有关键修改。后续本 session 中每次代码修改都必须继续追加到本文档，记录目标、设计判断、修改文件、验证方式和结果。
```
Comment (Flink-ddd) [inline docs/contributing/issue-108-session-log.md]: please use english to replace

File: rl_engine/kernels/gtest/op_checks.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        backend=candidate.backend,
+        total_outputs=total_outputs,
+        passed_outputs=passed_outputs,
+        pass_rate=pass_rate,
+        passed=passed_outputs == total_outputs,
+        cases=case_checks,
+    )
+
+
+def _run_case(
```
Comment (Flink-ddd) [inline rl_engine/kernels/gtest/op_checks.py:172]: Currently, _run_case only evaluates the forward pass (_call_candidate). As we have established in the RMSNorm, SwiGLU, and Embedding PRs, backward-pass consistency is critical to preventing RL training drift.

While it doesn't need to be implemented in this exact PR, you must add a TODO or open a tracking issue to support gradient checking in the gtest framework. Future iterations will need to set requires_grad=True on floating inputs, call .backward() on the outputs, and compare the resulting .grad tensors against the gold path using this same tolerance contract.

File: rl_engine/kernels/gtest/operator_specs.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    )
+
+
+def make_candidate(args: argparse.Namespace) -> CandidateSpec:
+    spec = OP_SPECS[args.op]
+    candidate_name = "pytorch" if args.candidate == "native" else args.candidate
+
+    if candidate_name in spec.candidate_paths:
+        candidate_op = _load_object(spec.candidate_paths[candidate_name])()
+        if args.op == "logp" and candidate_name == "cuda-sm90":
```
Comment (Flink-ddd) [inline rl_engine/kernels/gtest/operator_specs.py:100]: Injecting _LogpSM90CandidateAdapter using an if statement (if args.op == "logp" and candidate_name == "cuda-sm90":) inside make_candidate works for the MVP, but it will quickly become spaghetti code as you add more operators, ROCm backends, and Triton kernels that require shape-flattening or specific adapters.

Suggestion: Make the adapter mapping declarative. Add an optional candidate_adapters: dict[str, Callable] field to the OperatorSpec dataclass, so make_candidate can simply look up and wrap the candidate dynamically without needing to know operator-specific logic.

File: tests/test_logp.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        out = op.forward_fp32(logits, token_ids)
+        assert out.dtype == torch.float32
+
+    @pytest.mark.parametrize("dtype", [torch.float32, torch.bfloat16, torch.float16])
+    def test_forward_returns_input_dtype(self, dtype):
+        op = NativeLogpOp()
+        logits, token_ids = _make_inputs(2, 16, 257, dtype=dtype)
+        out = op.forward(logits, token_ids)
+        assert out.dtype == dtype
+
```
Comment (Flink-ddd) [inline tests/test_logp.py:52]: I noticed you included the unit tests for NativeLogpOp in this PR to prove the gtest framework works. Similar to the previous reference operators, test_logp.py is completely missing backward-pass tests (test_gradient_flows). Please ensure you add a slice-based Batch-Invariance (Axis-A) test for NativeLogpOp's backward pass before WS1 concludes.

File: docs/contributing/issue-108-session-log.md (hunk 1 above)
Comment (Flink-ddd) [reply docs/contributing/issue-108-session-log.md]: Please translate the session log from Chinese to English to maintain the repository's open-source linguistic consistency.

File: rl_engine/kernels/gtest/tolerance_contract.json [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        "float16": {"atol": 1.0e-3, "rtol": 1.0e-3}
+      },
+      "logprob": {
+        "float32": {"atol": 1.0e-5, "rtol": 0.0},
+        "bfloat16": {"atol": 5.0e-2, "rtol": 0.0},
+        "float16": {"atol": 5.0e-3, "rtol": 0.0}
+      }
+    },
+    "arch_overrides": {
+      "sm90": {}
```
Comment (Flink-ddd) [inline rl_engine/kernels/gtest/tolerance_contract.json:22]: Starting with sm90 as the only arch_overrides key is perfectly fine for this stage. The structure here is very clean, and it will be trivial to expand this to include rocm or cdna3 specific tolerances when you reach WS2 and WS3. No changes needed here.

**Review by Flink-ddd — CHANGES_REQUESTED** [review]
This is an excellently designed testing harness. Building a unified gtest runner with a YAML-backed tolerance contract is exactly what Workstream 1 needs to scale across CUDA, Triton, and ROCm. The JSON output integration will make CI regression tracking much easier.

Since this is an initial MVP version, the foundation looks very solid. Below are some reviews that outline some architectural improvements:

File: docs/contributing/issue-108-session-log.md (hunk 1 above)
Comment (frank-2077) [reply docs/contributing/issue-108-session-log.md]: I will fix it today

File: rl_engine/kernels/gtest/operator_specs.py (hunk 3 above)
Comment (frank-2077) [reply rl_engine/kernels/gtest/operator_specs.py:100]: Thinks for your comments. Our intended direction is that candiate wrappers should expose the same interface as the Pytorch gold path. The current SM90 adaoter is a temproary wrapper for SM90 CUDA interface, and I think we don't need any `if` judgement, we will delete the `if` in the future.
In current SM90 CUDA interface, the input shape is [batch*seq, vocba]. However, in the gtest, we don't flatten the batch, the shape is [batch, seq, vocba]

Comment (Flink-ddd) [reply rl_engine/kernels/gtest/operator_specs.py:100]: okay, make sense.

**Comment by Flink-ddd** [comment]
cc @maxiaosong1124 @inaniloquentee @KJLdefeated PTAL

File: rl_engine/kernels/gtest/op_checks.py (hunk 2 above)
Comment (frank-2077) [reply rl_engine/kernels/gtest/op_checks.py:172]: Thank very much for pointing that. I think it is critical for WS1, i will complete the backward part of gtest. 
In fact, I am doing it, iam using the [linear log-prob triton op](https://github.com/RL-Align/RL-Kernel/pull/122/changes#diff-7aa87ed6f985c1e9a073a5c97c9a612b9a3d0846bed111d53e754454e694e560) . It implemented forward and backward. I’ll polish my code tomorrow and submit an updated version.
image

File: rl_engine/kernels/gtest/op_checks.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        passed=all(output.passed for output in output_checks),
+        outputs=output_checks,
+    )
+
+
+# compatibility function or forward
+def _call_candidate(candidate: Callable[..., Any] | Any, inputs: Mapping[str, Any]) -> Any:
+    if hasattr(candidate, "forward") and callable(candidate.forward):
+        return candidate.forward(**inputs)
+    return candidate(**inputs)
```
Comment (KJLdefeated) [inline rl_engine/kernels/gtest/op_checks.py:296]: The harness builds inputs as {logits, token_ids} (matching NativeLogpOp) and dispatches with candidate(**inputs) (op_checks.py:184-187). But FusedLogpSM90Op.__call__(self, logits, labels) uses labels, not token_ids → TypeError. Only the explicit cuda-sm90 candidate gets the positional _LogpSM90CandidateAdapter; the registry candidate does not, so when the registry resolves to the SM90 op the kwargs don't match.

File: rl_engine/kernels/gtest/tolerance.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+
+_CONTRACT_PATH = Path(__file__).with_name("tolerance_contract.yaml")
+
+
+def load_contract(path: str | Path = _CONTRACT_PATH) -> dict[str, Any]:
+    """Load the dtype/operator-class tolerance contract."""
+
+    with Path(path).open("r", encoding="utf-8") as handle:
+        return json.load(handle)
```
Comment (KJLdefeated) [inline rl_engine/kernels/gtest/tolerance.py:18]: json.load on yaml files, this should be resolve.

**Review by KJLdefeated — CHANGES_REQUESTED** [review]

File: rl_engine/kernels/gtest/op_checks.py (hunk 2 above)
Comment (frank-2077) [reply rl_engine/kernels/gtest/op_checks.py:172]: Hi, @Flink-ddd  I have commit a new version for gtest. Add partial new code to support backward in gTest. E2E tests passed on both H20 and H100 GPUs. Note: tolerance tables may require further tuning. The figure shows an example of a backward test.
image

File: rl_engine/kernels/gtest/op_checks.py (hunk 6 above)
Comment (frank-2077) [reply rl_engine/kernels/gtest/op_checks.py:296]: Thanks for pointing that out, It made me realize that the registry may no longer be necessary. I’ll try removing it, so users only need to choose the backend they want to test.

File: rl_engine/kernels/gtest/tolerance.py (hunk 7 above)
Comment (frank-2077) [reply rl_engine/kernels/gtest/tolerance.py:18]: Good catch. The contract file is JSON-formatted but currently named .yaml, while the loader uses json.load. To avoid adding a YAML parser dependency and keep the file format explicit, I will rename it to tolerance_contract.json and update the loader/docs references accordingly.

**Comment by Flink-ddd** [comment]
please resolve the code conflicts first, then I will approve it. Thanks.

**Comment by frank-2077** [comment]
> please resolve the code conflicts first, then I will approve it. Thanks.

Fixed.

**Comment by Flink-ddd** [comment]
please resolve the CI error, Thanks.

**Review by Flink-ddd — APPROVED** [review]
LGTM, Thank you for update. cc @KJLdefeated PTAL, If there are no further issues, let's merge the first version of gtest. Thanks.

**Review by KJLdefeated — APPROVED** [review]
LGTM. Thx.
